osu.Game/Graphics/Containers/SectionsContainer.cs

```diff
@@ -177,36 +183,90 @@ protected override void UpdateAfterChildren()
+            headerBackgroundContainer.Y = ExpandableHeader?.Y ?? 0;
 
-                headerBackgroundContainer.Height = (ExpandableHeader?.LayoutSize.Y ?? 0) + (FixedHeader?.LayoutSize.Y ?? 0);
-                headerBackgroundContainer.Y = ExpandableHeader?.Y ?? 0;
+            selectMostVisible();
+        }
 
-                float scrollOffset = FixedHeader?.LayoutSize.Y ?? 0;
-                Func<T, float> diff = section => scrollContainer.GetChildPosInContent(section) - currentScroll - scrollOffset;
+        /// <summary>
+        /// Changes <see cref="SelectedSection"/> to currently most visible section
+        /// </summary>
+        private void selectMostVisible()
+        {
+            // reset last clicked section because user started scrolling themselves
+            if (scrollContainer.UserScrolling)
+                lastClickedSection = null;
 
-                if (scrollContainer.IsScrolledToEnd())
-                {
-                    SelectedSection.Value = Children.LastOrDefault();
-                }
-                else
-                {
-                    SelectedSection.Value = Children.TakeWhile(section => diff(section) <= 0).LastOrDefault()
-                                            ?? Children.FirstOrDefault();
-                }
+            // we are scrolled all the way to the bottom
+            // select the section user clicked for or the very last section
+            if (Precision.AlmostBigger(scrollContainer.Current, scrollContainer.ScrollableExtent))
+            {
+                SelectedSection.Value = lastClickedSection as T ?? Children.LastOrDefault();
+                return;
             }
+
+            var sortedByVisibility = new List<T>(Children);
+            sortedByVisibility.Sort(compareVisibility);
+            SelectedSection.Value = sortedByVisibility.FirstOrDefault();
+        }
+
+        private int compareVisibility(T a, T b)
+        {
+            // note: the decision to prioritize higher sections and differentiate scroll direction
+            // comes from the fact that our scroll target is the top of the screen
+            //
+            // if this is ever not the case, this comparison should be rewritten
+            float aIndex = IndexOf(a);
+            float bIndex = IndexOf(b);
+            float aVisibleHeight = visibleHeight(a);
+            float bVisibleHeight = visibleHeight(b);
+            // visibility requirement is lower (only 80%) for the section user clicked for
+            // this makes the section get selected sooner when scrolling upwards, to match it already getting selected sooner when scrolling downwards
+            bool aAllVisible = Precision.AlmostBigger(aVisibleHeight / a.Height, lastClickedSection as T == a ? 0.8f : 1.0f);
+            bool bAllVisible = Precision.AlmostBigger(bVisibleHeight / b.Height, lastClickedSection as T == b ? 0.8f : 1.0f);
```

Comment:
I still don't like this corner case and would ideally see it removed but /shrug.

Reply:
Like it says in the first comment, this differentiation between scroll directions is necessary because our scroll target is top of the screen. And it makes sense for it to be the top of the screen because most sections we scroll to have a title.